File: 백준/Gold/2294. 동전 2/동전 2.cc
#include<bits/stdc++.h>
using namespace std;
int n, k, dp[10001], coin, INF = 1e9;
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> n >> k;
    fill(dp, dp + 10001, INF);
    dp[0] = 0;
    for(int i = 0; i < n; i++){
        cin >> coin;
        if(coin >= 10001) continue;
        for(int j = coin; j <= k; j++){
            dp[j] = min(dp[j], dp[j - coin] + 1);
        }
    }
    if(dp[k] == INF) cout << -1 << '\n';
    else cout << dp[k] << '\n';
    return 0;
}

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    static final int INF = (int)1e9;
    static int[] dp = new int[10001];
    static int n, k, coin;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        k = sc.nextInt();
        Arrays.fill(dp, INF);
        dp[0] = 0;
        for (int i = 0; i < n; i++) {
            coin = sc.nextInt();
            if (coin >= 10001) continue;
            for (int j = coin; j <= k; j++) {
                dp[j] = Math.min(dp[j], dp[j - coin] + 1);
            }
        }
        if (dp[k] == INF) System.out.println(-1);
        else System.out.println(dp[k]);
    }
}
